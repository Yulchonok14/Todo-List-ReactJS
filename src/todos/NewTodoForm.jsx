import React, { useState } from 'react';

import styled from 'styled-components';

const NewTodoFormStyle = styled.div`
    border-radius: 8px;
    padding: 16px;
    text-align: center;
    box-shadow: 0 4px 8px grey;
`;

const NewTodoInput = styled.input`
    font-size: 16px;
    padding: 8px;
    border: none;
    border-bottom: 2px solid #ddd;
    border-radius: 8px;
    width: 70%;
    outline: none;
`;

const NewTodoButton = styled.button`
    font-size: 16px;
    padding: 8px;
    border: none;
    border-radius: 8px;
    outline: none;
    cursor: pointer;
    margin-left: 8px;
    width: 20%;
    background-color: #22ee22;
`;

const NewTodoForm = ({ todos, onCreateNewTodoClicked }) => {
  const [newTodo, setNewTodo] = useState('');

  return (
    <NewTodoFormStyle>
      <NewTodoInput
        placeholder="Type new todo item here"
        value={newTodo}
        onChange={(e) => setNewTodo(e.target.value)}
      />
      <NewTodoButton
        onClick={() => {
          const isDuplicated = todos.some((todo) => todo.text === newTodo);
          if (!isDuplicated) {
            onCreateNewTodoClicked(newTodo);
            setNewTodo('');
          }
        }}
      >
        Craft Todo
      </NewTodoButton>
    </NewTodoFormStyle>
  );
};

export default NewTodoForm;
