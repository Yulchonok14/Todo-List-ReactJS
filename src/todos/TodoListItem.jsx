import React from 'react';
import styled from 'styled-components';

const TodoItemContainer = styled.div`
    background: #fff;
    border-radius: 8px;
    margin-top: 8px;
    padding: 16px;
    position: relative;
    box-shadow: 0 4px 8px grey;
`;

export const getBorder = (startDate, today) => (
  startDate > new Date(today - 86400000 * 8) ? 'none' : '2px solid red'
);

const TodoItemContainerWithWarning = styled(TodoItemContainer)`
    border-bottom: ${(props) => getBorder(props.createdAt, Date.now())};
`;

const Button = styled.button`
    font-size: 16px;
    padding: 8px;
    border: none;
    border-radius: 8px;
    outline: none;
    cursor: pointer;
`;

const ButtonsContainer = styled.div`
    position: absolute;
    right: 12px;
    bottom: 12px;
`;

const RemoveButton = styled(Button)`
    display: inline-block;
    background-color: #ee2222;
    margin-left: 8px;
`;

const CompletedButton = styled(Button)`
    display: inline-block;
    background-color: #22ee22;
`;

const TodoListItem = ({ todo, onRemoveTodoClicked, onCompleteTodoClicked }) => {
  const Container = todo.isCompleted ? TodoItemContainer : TodoItemContainerWithWarning;
  return (
    <Container createdAt={todo.createdAt}>
      <h3>{todo.text}</h3>
      <p>
        Created at:&nbsp;
        {new Date(todo.createdAt).toLocaleDateString()}
      </p>
      <ButtonsContainer>
        {todo.isCompleted
          ? null
          : (
            <CompletedButton
              onClick={() => onCompleteTodoClicked(todo.id)}
            >
              Mark As Completed
            </CompletedButton>
          )}
        <RemoveButton
          onClick={() => onRemoveTodoClicked(todo.id)}
        >
          Remove
        </RemoveButton>
      </ButtonsContainer>
    </Container>
  );
};

export default TodoListItem;
