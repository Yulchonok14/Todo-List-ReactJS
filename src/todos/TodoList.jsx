import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import {
  addNewTodoRequest,
  completedTodoRequest,
  loadTodos,
  removeTodoRequest,
} from './thunks';
import {
  getCompletedTodos,
  getIncompletedTodos,
  getLoading,
  getTodos,
} from './selectors';

import NewTodoForm from './NewTodoForm';
import TodoListItem from './TodoListItem';

const ListWrapper = styled.div`
    max-width: 700px;
    margin: auto;
`;

const TodoList = ({
  incompletedTodos,
  completedTodos,
  isLoading,
  onRemoveTodoClicked,
  onCompleteTodoClicked,
  onLoadTodos,
  onCreateNewTodoClicked,
  todos,
}) => {
  useEffect(() => {
    onLoadTodos();
  }, []);

  const loading = <div>Loading todos...</div>;
  const content = (
    <ListWrapper>
      <NewTodoForm
        todos={todos}
        onCreateNewTodoClicked={onCreateNewTodoClicked}
      />
      <h3 className="title incomplete">Incompleted:</h3>
      {incompletedTodos.map((todo) => (
        <TodoListItem
          key={todo.id}
          onCompleteTodoClicked={onCompleteTodoClicked}
          onRemoveTodoClicked={onRemoveTodoClicked}
          todo={todo}
        />
      ))}
      <h3 className="title complete">Completed:</h3>
      {completedTodos.map((todo) => (
        <TodoListItem
          key={todo.id}
          onCompleteTodoClicked={onCompleteTodoClicked}
          onRemoveTodoClicked={onRemoveTodoClicked}
          todo={todo}
        />
      ))}
    </ListWrapper>
  );
  return isLoading ? loading : content;
};

const mapStateToProps = (state) => ({
  incompletedTodos: getIncompletedTodos(state),
  completedTodos: getCompletedTodos(state),
  isLoading: getLoading(state),
  todos: getTodos(state),
});

const mapDispatchToProps = (dispatch) => ({
  onRemoveTodoClicked: (id) => dispatch(removeTodoRequest(id)),
  onCompleteTodoClicked: (id) => dispatch(completedTodoRequest(id)),
  onLoadTodos: () => dispatch(loadTodos()),
  onCreateNewTodoClicked: (text) => dispatch(addNewTodoRequest(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(TodoList);

export { TodoList };
