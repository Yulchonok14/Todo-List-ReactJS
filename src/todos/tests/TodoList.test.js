/* eslint-disable react/jsx-filename-extension */
/* eslint-env mocha */
import 'jsdom-global/register';
import chai, { expect } from 'chai';
import React from 'react';
import chaiJestSnapshot from 'chai-jest-snapshot';
import renderer from 'react-test-renderer';
import { configure, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import sinon from 'sinon';
import { TodoList } from '../TodoList';
import TodoListItem from '../TodoListItem';
import NewTodoForm from '../NewTodoForm';

chai.use(chaiJestSnapshot);

configure({
  adapter: new Adapter(),
});

before(() => {
  chaiJestSnapshot.resetSnapshotRegistry();
});

// eslint-disable-next-line func-names
beforeEach(function () {
  chaiJestSnapshot.configureUsingMochaContext(this);
});

describe('TodoList', () => {
  it('renders correctly', () => {
    const mockOnLoadTodos = sinon.stub();
    const mockIncompletedTodos = [{
      id: 1,
      text: 'Go to the cinema',
      isCompleted: false,
    },
    {
      id: 2,
      text: 'Go to the theater',
      isCompleted: false,
    }];
    const mockCompletedTodos = [{
      id: 3,
      text: 'Go to the skating ring',
      isCompleted: true,
    },
    {
      id: 4,
      text: 'Meet with friends',
      isCompleted: true,
    }];
    const component = renderer.create(<TodoList
      incompletedTodos={mockIncompletedTodos}
      completedTodos={mockCompletedTodos}
      onLoadTodos={mockOnLoadTodos}
    />);
    const tree = component.toJSON();

    expect(tree).to.matchSnapshot();
  });

  it('has form, headings and 2 set of items', () => {
    const mockOnRemoveTodoCallBack = sinon.stub();
    const mockOnCompleteTodoCallBack = sinon.stub();
    const mockOnCreateNewTodoCallBack = sinon.stub();
    const mockOnLoadTodos = sinon.stub();
    const mockIncompletedTodos = [{
      id: 1,
      text: 'Go to the cinema',
      isCompleted: false,
    },
    {
      id: 2,
      text: 'Go to the theater',
      isCompleted: false,
    }];
    const mockCompletedTodos = [{
      id: 3,
      text: 'Go to the skating ring',
      isCompleted: true,
    },
    {
      id: 4,
      text: 'Meet with friends',
      isCompleted: true,
    }];
    const wrapper = mount(<TodoList
      onCreateNewTodoClicked={mockOnCreateNewTodoCallBack}
      onRemoveTodoClicked={mockOnRemoveTodoCallBack}
      onCompleteTodoClicked={mockOnCompleteTodoCallBack}
      incompletedTodos={mockIncompletedTodos}
      completedTodos={mockCompletedTodos}
      onLoadTodos={mockOnLoadTodos}
      isLoading={false}
    />);

    expect(wrapper.find(NewTodoForm)).to.have.lengthOf(1);

    expect(wrapper.find('h3.title')).to.have.length(2);
    expect(wrapper.find('h3.title').at(0).text().includes('Incompleted:')).to.equal(true);
    expect(wrapper.find('h3.title').at(1).text().includes('Completed:')).to.equal(true);

    expect(wrapper.find(TodoListItem)).to.have.lengthOf(4);
    const completedItems = wrapper.find('h3.title.complete ~ *').filterWhere((n) => n.type() === TodoListItem);
    expect(completedItems).to.have.lengthOf(2);
  });

  it('shows loading text while loading', () => {
    const mockOnRemoveTodoCallBack = sinon.stub();
    const mockOnCompleteTodoCallBack = sinon.stub();
    const mockOnCreateNewTodoCallBack = sinon.stub();
    const mockOnLoadTodos = sinon.stub();
    const wrapper = mount(<TodoList
      onCreateNewTodoClicked={mockOnCreateNewTodoCallBack}
      onRemoveTodoClicked={mockOnRemoveTodoCallBack}
      onCompleteTodoClicked={mockOnCompleteTodoCallBack}
      incompletedTodos={[]}
      completedTodos={[]}
      onLoadTodos={mockOnLoadTodos}
      isLoading
    />);

    expect(wrapper.find('div').text()).to.be.equal('Loading todos...');
    expect(wrapper.find(TodoListItem)).to.have.lengthOf(0);
  });
});
