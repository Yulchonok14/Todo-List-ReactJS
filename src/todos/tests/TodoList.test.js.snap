// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders correctly 1`] = `
<div
  className="sc-pGacB hvzgvV"
>
  <div
    className="sc-bdfBQB hpKxwM"
  >
    <input
      className="sc-gsTEea NRtLr"
      onChange={[Function]}
      placeholder="Type new todo item here"
      value=""
    />
    <button
      className="sc-dlfnuX eOyWug"
      onClick={[Function]}
    >
      Craft Todo
    </button>
  </div>
  <h3
    className="title incomplete"
  >
    Incompleted:
  </h3>
  <div
    className="sc-hKgJUU sc-eCstlR kMigaP hlAxGi"
  >
    <h3>
      Go to the cinema
    </h3>
    <p>
      Created at: 
      Invalid Date
    </p>
    <div
      className="sc-gKseQn jwGGfw"
    >
      <button
        className="sc-jSgvzq sc-fubCzh dFYixR llQoPR"
        onClick={[Function]}
      >
        Mark As Completed
      </button>
      <button
        className="sc-jSgvzq sc-iBPTik dFYixR bfvxIP"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </div>
  <div
    className="sc-hKgJUU sc-eCstlR kMigaP hlAxGi"
  >
    <h3>
      Go to the theater
    </h3>
    <p>
      Created at: 
      Invalid Date
    </p>
    <div
      className="sc-gKseQn jwGGfw"
    >
      <button
        className="sc-jSgvzq sc-fubCzh dFYixR llQoPR"
        onClick={[Function]}
      >
        Mark As Completed
      </button>
      <button
        className="sc-jSgvzq sc-iBPTik dFYixR bfvxIP"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </div>
  <h3
    className="title complete"
  >
    Completed:
  </h3>
  <div
    className="sc-hKgJUU kMigaP"
  >
    <h3>
      Go to the skating ring
    </h3>
    <p>
      Created at: 
      Invalid Date
    </p>
    <div
      className="sc-gKseQn jwGGfw"
    >
      <button
        className="sc-jSgvzq sc-iBPTik dFYixR bfvxIP"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </div>
  <div
    className="sc-hKgJUU kMigaP"
  >
    <h3>
      Meet with friends
    </h3>
    <p>
      Created at: 
      Invalid Date
    </p>
    <div
      className="sc-gKseQn jwGGfw"
    >
      <button
        className="sc-jSgvzq sc-iBPTik dFYixR bfvxIP"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </div>
</div>
`;
